base/compiler/abstractinterpretation.jl

```diff
@@ -349,52 +354,41 @@ function abstract_call_method(interp::AbstractInterpreter, method::Method, @nosp
-            if topmost === nothing && method2 === inf_method2
-                if hardlimit
-                    topmost = infstate
-                    edgecycle = true
-                else
+            if callee_method2 === inf_method2
+                if !hardlimit
                     # if this is a soft limit,
                     # also inspect the parent of this edge,
                     # to see if they are the same Method as sv
                     # in which case we'll need to ensure it is convergent
                     # otherwise, we don't
-                    for parent in infstate.callers_in_cycle
-                        # check in the cycle list first
-                        # all items in here are mutual parents of all others
-                        parent_method2 = parent.src.method_for_inference_limit_heuristics # limit only if user token match
-                        parent_method2 isa Method || (parent_method2 = nothing) # Union{Method, Nothing}
-                        if parent.linfo.def === sv.linfo.def && sv_method2 === parent_method2
-                            topmost = infstate
-                            edgecycle = true
-                            break
+
+                    # check in the cycle list first
+                    # all items in here are mutual parents of all others
+                    if !_any(matches_sv, infstate.callers_in_cycle)
+                        let parent = infstate.parent
+                            parent !== nothing || continue
+                            parent = parent::InferenceState
+                            (parent.cached || parent.parent !== nothing) || continue
+                            matches_sv(parent) || continue
                         end
                     end
-                    let parent = infstate.parent
-                        # then check the parent link
-                        if topmost === nothing && parent !== nothing
-                            parent = parent::InferenceState
-                            parent_method2 = parent.src.method_for_inference_limit_heuristics # limit only if user token match
-                            parent_method2 isa Method || (parent_method2 = nothing) # Union{Method, Nothing}
-                            if (parent.cached || parent.parent !== nothing) && parent.linfo.def === sv.linfo.def && sv_method2 === parent_method2
-                                topmost = infstate
-                                edgecycle = true
-                            end
+
+                    # If the method defines a recursion relation, give it a chance
+                    # to tell us that this recursion is actually ok.
+                    if isdefined(method, :recursion_relation)
+                        if Core._apply_pure(method.recursion_relation, Any[method, callee_method2, sig, infstate.linfo.specTypes])
+                            continue
```

Comment:
Per your comments about transitivity, I think this might need to check every frame for which the methods are the same (prior to `topmost === nothing || continue` guard). But I'll discuss more offline with you.